manager/controlapi/secret.go

```diff
@@ -12,13 +17,40 @@ import (
 // MaxSecretSize is the maximum byte length of the `Secret.Spec.Data` field.
 const MaxSecretSize = 500 * 1024 // 500KB
 
+var isValidSecretName = regexp.MustCompile(`^[a-zA-Z0-9](?:[a-zA-Z0-9-_.]{0,62}[a-zA-Z0-9])*$`)
+
+// assumes spec is not nil
+func secretFromSecretSpec(spec *api.SecretSpec) *api.Secret {
+	return &api.Secret{
+		ID:         identity.NewID(),
+		Spec:       *spec,
+		SecretSize: uint32(len(spec.Data)),
+		Digest:     digest.FromBytes(spec.Data).String(),
```

Comment:
Out of band discussion:

using:  `string digest = 4  [(gogoproto.customtype) = "github.com/docker/distribution/digest.Digest"];`, `make generate` produces a compiled `object.pb.go` file with no errors, but that file refuses to build:

```
$  make build
🐳 build
github.com/docker/swarmkit
github.com/docker/swarmkit/api/duration
github.com/docker/swarmkit/api/timestamp
github.com/docker/swarmkit/protobuf/plugin
github.com/docker/swarmkit/manager/testcluster
github.com/docker/swarmkit/protobuf/plugin/deepcopy/test
github.com/docker/swarmkit/protobuf/plugin/raftproxy/test
github.com/docker/swarmkit/protobuf/ptypes
github.com/docker/swarmkit/api
github.com/docker/swarmkit/protobuf/plugin/authenticatedwrapper
github.com/docker/swarmkit/protobuf/plugin/deepcopy
github.com/docker/swarmkit/cmd/protoc-gen-gogoswarm
# github.com/docker/swarmkit/api
api/objects.pb.go:1307: invalid argument m.Digest (type *digest.Digest) for len
api/objects.pb.go:1310: invalid argument m.Digest (type *digest.Digest) for len
api/objects.pb.go:1311: second argument to copy should be slice or string; have *digest.Digest
api/objects.pb.go:1597: invalid argument m.Digest (type *digest.Digest) for len
api/objects.pb.go:3844: cannot use digest.Digest(data[iNdEx:postIndex]) (type digest.Digest) as type *digest.Digest in assignment
make: *** [build] Error 2
```

This may be a bug with `gogo`, so we're just going to leave it a string for now and maybe work it out later.